File: typescript/triangle/triangle.ts
export default class Triangle {
  sides: number[];

  constructor(...sides: number[]) {
    this.sides = sides;
  }

  public checkValidity() {
    const [side1, side2, side3] = this.sides;
    const hasZeroSide = side1 === 0 && side2 === 0 && side3 === 0;
    const hasNegativeSide = side1 < 0 || side2 < 0 || side3 < 0;

    if (hasZeroSide || hasNegativeSide || !this.hasValidDimensions()) {
      throw new Error("Invalid triangle dimensions");
    }
  }

  public hasValidDimensions() {
    const [shortestSide, middleSide, longestSide] = this.sides.sort(
      (a, b) => a - b
    );
    return shortestSide + middleSide >= longestSide;
  }

  public isEquilateral() {
    const [side1, side2, side3] = this.sides;
    return side1 === side2 && side2 === side3;
  }

  public isIsosceles() {
    const [side1, side2, side3] = this.sides;
    return (
      !this.isEquilateral() &&
      (side1 === side2 || side1 === side3 || side2 === side3)
    );
  }

  public isScalene() {
    return !this.isEquilateral() && !this.isIsosceles();
  }

  kind() {
    this.checkValidity();

    if (this.isEquilateral()) return "equilateral";
    if (this.isIsosceles()) return "isosceles";
    if (this.isScalene()) return "scalene";
  }
}

File: kotlin/triangle/src/main/kotlin/Triangle.kt
class Triangle<out T : Number>(a: T, b: T, c: T) {
    init {
        require(a.toDouble() < b.toDouble() + c.toDouble())
        require(b.toDouble() < a.toDouble() + c.toDouble())
        require(c.toDouble() < a.toDouble() + b.toDouble())
    }

    private val uniqueSides = setOf(a, b, c).size

    val isEquilateral: Boolean = uniqueSides == 1
    val isIsosceles: Boolean =  uniqueSides <= 2
    val isScalene: Boolean = uniqueSides == 3
}
